### PR TITLE
[DEV-6439] Publish dates endpoint

### PR DESCRIPTION
**Pull request by emily-brents**
**Description:**
Creates `api/v2/reporting/agencies/publish_dates/ endpoint, which provides data about agencies submissions for a provided fiscal year.

**Technical details:**
Queries the submisssion_attributes table for most of the data, joins to toptier_agency and reporting_agency_overview. Sorting is being done in python due to the publication_date sort having to select a specific period and then sort by the publication date of that period.

**Requirements for PR merge:**

1. [X] Unit & integration tests updated
2. [X] API documentation updated
3. [x] Necessary PR reviewers:
    - [x] Backend
    - [x] Frontend 
4. [N/A] Matview impact assessment completed
5. [X] Frontend impact assessment completed
6. [X] Data validation completed
7. [N/A] Appropriate Operations ticket(s) created
8. [X] Jira Ticket [DEV-6439](https://federal-spending-transparency.atlassian.net/browse/DEV-6439):
    - [X] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [ ] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

File: usaspending_api/reporting/v2/views/agencies/publish_dates.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+from usaspending_api.submissions.models import SubmissionAttributes
+
+
+class PublishDates(AgencyBase, PaginationMixin):
+    """Returns list of agency submission information, included published and certified dates for the fiscal year"""
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/v2/reporting/agencies/publish_dates.md"
+
+    def validate_publication_sort(self, sort_key):
+        regex_string = r"publication_date,([1-9]|1[0-2])"
```
Comment (brett-varney) [inline usaspending_api/reporting/v2/views/agencies/publish_dates.py:30]: shouldn't that be [2-9]|1[0-2]  ?

**Review by brett-varney — COMMENTED** [review]
initial re thought

File: usaspending_api/reporting/v2/views/agencies/publish_dates.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            {"page_metadata": page_metadata, "results": results, "messages": self.standard_response_messages}
+        )
+
+    @cached_property
+    def pagination(self):
+        sortable_columns = [
+            "agency_name",
+            "abbreviation",
+            "agency_code",
+            "current_total_budget_authority_amount",
```
Comment (brett-varney) [inline usaspending_api/reporting/v2/views/agencies/publish_dates.py:126]: Contract shows "name" and "code" as sortable; does this code or the contract need to change?

**Review by brett-varney — COMMENTED** [review]
another question

File: usaspending_api/reporting/v2/views/agencies/publish_dates.py (hunk 2 above)
Comment (brett-varney) [reply usaspending_api/reporting/v2/views/agencies/publish_dates.py:126]: My mistake; old version

Comment (emily-brents) [reply usaspending_api/reporting/v2/views/agencies/publish_dates.py:126]: The contract lists the fields as `agency_code` and `agency_name`

File: usaspending_api/reporting/v2/views/agencies/publish_dates.py (hunk 1 above)
Comment (emily-brents) [reply usaspending_api/reporting/v2/views/agencies/publish_dates.py:30]: That would be correct.

File: usaspending_api/reporting/v2/views/agencies/publish_dates.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            raise UnprocessableEntityException(
+                "publication_date sort param must be in the format 'publication_date,<fiscal_period>'"
+            )
+
+    def get_agency_data(self):
+        agency_filters = [Q(toptier_code=OuterRef("toptier_code"))]
+        if self.filter is not None:
+            agency_filters.append(Q(name__icontains=self.filter))
+        results = (
+            SubmissionAttributes.objects.filter(reporting_fiscal_year=self.fiscal_year)
```
Comment (tony-sappe) [inline usaspending_api/reporting/v2/views/agencies/publish_dates.py:41]: This appears to only be obtaining agencies based on the submissions for the specific year (it isn't even filtering out published submissions before a submission window closes which should be an implicit expectation for all account data being returning by the API at this point). See comments in the Jira ticket for a little more detail, specifically https://federal-spending-transparency.atlassian.net/browse/DEV-6439?focusedCommentId=63809

Comment (emily-brents) [reply usaspending_api/reporting/v2/views/agencies/publish_dates.py]: yup, this should be addressed now. I've switched the query to use ToptierAgencies.objects.account_agencies, so it'll return all agencies with account data for any fiscal year, as well as all periods w/ blanks if there is no submission for that period.

**Review by tony-sappe — APPROVED** [review]

**Pull request merged**